tower-balance/src/p2c/service.rs

```diff
@@ -253,79 +165,41 @@ where
         fn(<D::Service as Service<Req>>::Error) -> error::Error,
     >;
 
-    /// Prepares the balancer to process a request.
-    ///
-    /// When `Async::Ready` is returned, `ready_index` is set with a valid index
-    /// into `ready` referring to a `Service` that is ready to disptach a request.
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        // First and foremost, process discovery updates. This removes or updates a
-        // previously-selected `ready_index` if appropriate.
-        self.poll_discover()?;
-
-        // Drive new or busy services to readiness.
-        self.poll_unready();
-        trace!({ nready = self.ready_services.len(), nunready = self.unready_services.len() }, "poll_ready");
-
+        self.update_pending_from_discover()?;
+        self.promote_pending_to_ready();
         loop {
-            // If a node has already been selected, ensure that it is ready.
+            // If a service has already been selected, ensure that it is ready.
             // This ensures that the underlying service is ready immediately
             // before a request is dispatched to it. If, e.g., a failure
             // detector has changed the state of the service, it may be evicted
-            // from the ready set so that P2C can be performed again.
-            if let Some(index) = self.next_ready_index {
-                trace!({ next.idx = index }, "preselected ready_index");
-                debug_assert!(index < self.ready_services.len());
-
-                if let Ok(Async::Ready(())) = self.poll_ready_index_or_evict(index) {
-                    return Ok(Async::Ready(()));
+            // from the ready set so that another service can be selected.
+            if let Some(index) = self.ready_index.take() {
+                match self.services.check_ready_index(index) {
```

Comment:
This "extra" ready check is still so weird to me, and I worry that it might have adverse consequences as calling `poll_ready` now needs to be idempotent. Maybe that's fine, I'm just not sure how much the extra check buys us..

Reply:
I see and follow the argument made in the API docs for `ReadyCache` for why you need to check ready again, it's just a weird contract to me. You always run the risk of the service _just_ becoming not ready between the last time you called `poll_ready` and `call`. Does calling `poll_ready` again just before `call` serve that much of a purpose? And is it worth the cost of calling `poll_ready` again?

Reply:
> `poll_ready` now needs to be idempotent.

This is already the case! You already have to expect that `poll_ready` may be called redundantly -- consider a Future that needs multiple services to be ready before it can take an action: there will almost certainly be redundant calls:
```rust
try_ready!(a.poll_ready())
try_ready!(b.poll_ready())
a.call(c).join(b.call(d))
```

> You always run the risk of the service just becoming not ready between the last time you called poll_ready and call

Sure. There will always be the potential for a resource to become unavailable just as its used. It doesn't mean that we shouldn't try to minimize that potential, though.

Specifically, we have decent guarantees that the thread has not changed/reactor state between poll_ready and call because we know the task has not been interrupted! If we poll_ready, and then let the thread/reactor do additional processing before call, it seems _much_ more likely for the state to change. The readiness contract is: given the most up-to-date state of the runtime, can we proceed? That state _can't_ change between poll_ready and call if they run in a single task invocation, really. (Yes, the external world can change, but we're operating under the best guarantees the local task can offer us).

> Does calling poll_ready again just before call serve that much of a purpose? And is it worth the cost of calling poll_ready again?

This depends entirely on the cost of calling `poll_ready`. I posit that implementors _should_ make `poll_ready` cheap to call, especially if it's already in a ready state. As discussed above, I think this guidance holds regardless of the semantics of ReadyCache.

--

On further thought, it is not _required_ that a ready service be checked before call is invoked. But it is _necessary_ (for our needs in the balancer, at least) to have a means to check a ready service redundantly. The balancer, specifically, could have services that become ready and remain unused for _minutes_. It's absolutely necessary to have means to re-check for readiness in this case.

Reply:
Okay, I think that answer persuades me. Maybe we could add a fast-path later that doesn't call `poll_ready` again if we haven't been interrupted since the service was last discovered to be ready, but let's not engage in premature optimization here.

Reply:
I'm going to update the docs to reflect a more relaxed contract where calls to `check_ready*`  are optional; though we should continue to use it from the Balancer...